src/NuGet.Clients/PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs

```diff
@@ -306,34 +311,62 @@ private Visibility EvaluateStatusBarVisibility(IItemLoader<PackageItemListViewMo
 
         private void UpdatePackageList(IEnumerable<PackageItemListViewModel> packages, bool refresh)
         {
-            // remove the loading status indicator if it's in the list
-            Items.Remove(_loadingStatusIndicator);
-
-            if (refresh)
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
             {
-                ClearPackageList();
-            }
+                // Synchronize updating Items list
+                await _itemsLock.WaitAsync();
 
-            // add newly loaded items
-            foreach (var package in packages)
-            {
-                package.PropertyChanged += Package_PropertyChanged;
-                Items.Add(package);
-                _selectedCount = package.Selected ? _selectedCount + 1 : _selectedCount;
-            }
+                try
+                {
+                    // remove the loading status indicator if it's in the list
+                    Items.Remove(_loadingStatusIndicator);
+
+                    if (refresh)
+                    {
+                        ClearPackageList();
+                    }
+
+                    // add newly loaded items
+                    foreach (var package in packages)
+                    {
+                        package.PropertyChanged += Package_PropertyChanged;
+                        Items.Add(package);
+                        _selectedCount = package.Selected ? _selectedCount + 1 : _selectedCount;
+                    }
+
+                    Items.Add(_loadingStatusIndicator);
+                }
 
-            Items.Add(_loadingStatusIndicator);
+                finally
+                {
+                    _itemsLock.Release();
+                }
+            });
         }
 
         private void ClearPackageList()
         {
-            foreach (var package in PackageItems)
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
```

Comment:
It does run on UI thread, and as part of loadItems api. As discussed, I'll use Wait here instead of WaitAsync